crypto/s390x_arch.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef S390X_ARCH_H
+# define S390X_ARCH_H
+
+/*
+ * The elements of OPENSSL_s390xcap_P are the doublewords returned by the STFLE
+ * instruction followed by the doubleword pairs returned by instructions' QUERY
+ * functions. If STFLE returns fewer doublewords or an instruction is not
+ * supported, the corresponding element is zero. The order is as follows:
+ *
+ * STFLE:STFLE:STFLE.
+ * KIMD:KIMD:KLMD:KLMD:KM:KM:KMC:KMC:KMAC:KMAC:KMCTR:KMCTR:KMO:KMO:KMF:KMF
+ * :PRNO:PRNO:KMA:KMA
+ */
+# define S390X_STFLE_DWORDS	3
+# define S390X_QUERY_DWORDS	20
+# define S390X_CAP_DWORDS	(S390X_STFLE_DWORDS + S390X_QUERY_DWORDS)
+extern unsigned long long OPENSSL_s390xcap_P[];
+
+/* Offsets */
+# define S390X_STFLE		0
+# define S390X_KIMD		(S390X_STFLE + S390X_STFLE_DWORDS)
+# define S390X_KLMD		(S390X_KIMD + 2)
+# define S390X_KM		(S390X_KLMD + 2)
+# define S390X_KMC		(S390X_KM + 2)
+# define S390X_KMAC		(S390X_KMC + 2)
+# define S390X_KMCTR		(S390X_KMAC + 2)
+# define S390X_KMO		(S390X_KMCTR + 2)
+# define S390X_KMF		(S390X_KMO + 2)
+# define S390X_PRNO		(S390X_KMF + 2)
+# define S390X_KMA		(S390X_PRNO + 2)
+
+/* OPENSSL_s390xcap_P[S390X_STFLE + 2] flags */
+# define S390X_STFLE_VXE	(1ULL << 56)
+# define S390X_STFLE_VXD	(1ULL << 57)
+# define S390X_STFLE_VX		(1ULL << 62)
+
+/* OPENSSL_s390xcap_P[S390X_KIMD] flags */
+# define S390X_KIMD_SHAKE_256	(1ULL << 26)
+# define S390X_KIMD_SHAKE_128	(1ULL << 27)
+# define S390X_KIMD_SHA3_512	(1ULL << 28)
+# define S390X_KIMD_SHA3_384	(1ULL << 29)
+# define S390X_KIMD_SHA3_256	(1ULL << 30)
+# define S390X_KIMD_SHA3_224	(1ULL << 31)
+
+/* OPENSSL_s390xcap_P[S390X_KLMD] flags */
+# define S390X_KLMD_SHAKE_256	(1ULL << 26)
```

Comment:
It's not really coincidence that corresponding flags are same in both KIMD and KLMD cases (as well as in KM and KMC), so why duplicate? Wouldn't defining S390X_SHAKE_256 be sufficient as claiming that it applies in both cases? Alternatively one can #define S390X_KLMD_SHAKE_256 S390X_KIMD_SHAKE_256. [This naturally applies to all similar cases, including KM\*.]

Reply:
I just made one define per instruction function code. Id like to keep the mnemonic in the define name so its obvious at which offset (or struct field) this flag is to be used. Ill implement your second option.